File: src/graph/dot.rs
use {
    super::CssClass,
    crate::graph::{Cell, Edge, Subgraph, TableNode},
    std::iter,
};

pub(crate) fn escape_html(s: &str) -> String {
    s.replace("&", "&amp;")
        .replace("\"", "&quot;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
}

pub(crate) struct Dot;

impl Dot {
    pub fn generate_dot_source<T, E>(
        tables: T,
        // nodes: &[Node],
        edges: E,
        subgraphs: &[Subgraph],
    ) -> String
    where
        T: Iterator<Item = TableNode>,
        E: Iterator<Item = Edge>,
    {
        let tables = tables
            .map(|table| {
                format!(
                    r#"
    "{id}" [id="{id}", label=<
        <TABLE BORDER="0" CELLBORDER="1" CELLSPACING="8" CELLPADDING="4">
        <TR><TD WIDTH="230" BORDER="0" CELLPADDING="6" HREF="remove_me_url.title">{title}</TD></TR>
        {sections}
        <TR><TD BORDER="0"></TD></TR>
        </TABLE>
    >];
                    "#,
                    id = table.id,
                    title = table.title,
                    sections = table
                        .sections
                        .iter()
                        .map(|node| Dot::process_cell(table.id, node))
                        .collect::<Vec<_>>()
                        .join("\n"),
                )
            })
            .collect::<Vec<_>>()
            .join("\n");

        format!(
            r#"
digraph {{
    graph [
        rankdir = "LR"
        ranksep = 2.0
        fontname = "Arial"
    ];
    node [
        fontsize = "16"
        fontname = "Arial"
        shape = "plaintext"
        style = "rounded, filled"
    ];
    edge [
        label = " "
    ];

    {}

    {}

    {}
}}
            "#,
            tables,
            Dot::clusters(subgraphs),
            Dot::process_edges(edges),
        )
    }

    fn process_cell(table_id: u32, cell: &Cell) -> String {
        let styles = [
            cell.style
                .border
                .map_or(String::new(), |b| format!(r#"BORDER="{}""#, b)),
            cell.style
                .rounded
                .then_some(r#"STYLE="ROUNDED""#.to_string())
                .unwrap_or(String::new()),
        ]
        .join(" ");

        let port = format!("{}_{}", cell.range_start.0, cell.range_start.1);

        if cell.children.is_empty() {
            format!(
                r#"     <TR><TD PORT="{port}" ID="{table_id}:{port}" {styles} {href}>{i}{name}</TD></TR>"#,
                href = Dot::css_classes_href(&cell.style.classes),
                i = cell
                    .style
                    .icon
                    .map(|c| format!("<B>{}</B>  ", c))
                    .unwrap_or(String::new()),
                name = escape_html(&cell.title),
            )
        } else {
            let (cell_styles, table_styles) = (r#"BORDER="0""#.to_string(), styles);

            let dot_cell = format!(
                r#"     <TR><TD PORT="{port}" {cell_styles} {href}>{name}</TD></TR>"#,
                href = String::new(),
                name = escape_html(&cell.title),
            );

            format!(
                r#"
            <TR><TD BORDER="0" CELLPADDING="0">
            <TABLE ID="{table_id}:{port}" CELLSPACING="4" CELLPADDING="4" CELLBORDER="1" {table_styles} BGCOLOR="green" {href}>
            {}
            </TABLE>
            </TD></TR>
            "#,
                iter::once(dot_cell)
                    .chain(
                        cell.children
                            .iter()
                            .map(|item| Dot::process_cell(table_id, item))
                    )
                    .collect::<Vec<_>>()
                    .join("\n"),
                href = Dot::css_classes_href(&cell.style.classes),
            )
        }
    }

    fn process_edges<E>(edges: E) -> String
    where
        E: Iterator<Item = Edge>,
    {
        edges
            .map(|edge| {
                let from = format!(r#"{}:"{}_{}""#, edge.from.0, edge.from.1, edge.from.2);
                let to = format!(r#"{}:"{}_{}""#, edge.to.0, edge.to.1, edge.to.2);

                let attrs = iter::once(format!(
                    r#"id="{}:{}_{} -> {}:{}_{}""#,
                    edge.from.0, edge.from.1, edge.from.2, edge.to.0, edge.to.1, edge.to.2,
                ))
                .chain(iter::once(Dot::css_classes_href(&edge.styles)))
                .filter(|s| !s.is_empty())
                .collect::<Vec<_>>();

                format!("{} -> {} [{attrs}];", from, to, attrs = attrs.join(", "),)
            })
            .collect::<Vec<_>>()
            .join("\n    ")
    }

    fn clusters(subgraphs: &[Subgraph]) -> String {
        subgraphs
            .iter()
            .map(|subgraph| {
                format!(
                    r#"
        subgraph "cluster_{name}" {{
            label = "{name}";

            {nodes}

            {subgraph}
        }};
                    "#,
                    name = subgraph.title,
                    nodes = subgraph.nodes.join(" "),
                    subgraph = Dot::clusters(&subgraph.subgraphs),
                )
            })
            .collect::<Vec<_>>()
            .join("\n")
    }

    fn css_classes_href(classes: &[CssClass]) -> String {
        if classes.is_empty() {
            "".to_string()
        } else {
            format!(
                r#"href="remove_me_url.{}""#,
                classes
                    .iter()
                    .map(|c| c.to_str())
                    .collect::<Vec<_>>()
                    .join(".")
            )
        }
    }
}

File: editors/code/src/dotGenerator/dot.ts

import { Cell, CssClass, Edge, Subgraph, TableNode } from './graph';

function escapeHtml(s: string): string {
  return s.replace(/&/g, "&amp;")
    .replace(/"/g, "&quot;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

const EMPTY_STRING = "";

//      .filter(node => !node.isCollapsed)
export class Dot {
  static generateDotSourceString(tables: TableNode[], edges: Edge[], subgraphs: Subgraph[]): string {
    const tablesStr = tables.map(table => {
        return `
"${table.id}" [id="${table.id}", label=<
    <TABLE BORDER="0" CELLBORDER="1" CELLSPACING="8" CELLPADDING="4">
    <TR><TD WIDTH="230" BORDER="0" CELLPADDING="6" HREF="remove_me_url.title">${table.title}</TD></TR>
    ${table.sections

      .map(node => Dot.processCell(table.id, node)).join("\n")}
    <TR><TD CELLSPACING="0" HEIGHT="1" WIDTH="1" FIXEDSIZE="TRUE" STYLE="invis"></TD></TR>
    </TABLE>
>];
`     ;}).join("\n");

    return `
digraph {
  graph [
    rankdir = "LR"
    ranksep = 2.0
    fontname = "Arial"
  ];
  node [
    fontsize = "16"
    fontname = "Arial"
    shape = "plaintext"
    style = "rounded, filled"
  ];
  edge [
    label = " "
  ];

  ${tablesStr}

  ${Dot.clusters(subgraphs)}

  ${Dot.processEdges(edges)}
}
    `;
  }

  static processCell(tableId: number, cell: Cell): string {
    const styles = [
      cell.style.border ? `BORDER="${cell.style.border}"` : "",
      cell.style.rounded ? `STYLE="ROUNDED"` : ""
    ].join(" ");

    const title = `${cell.style.icon ? `<B>${cell.style.icon}</B>  ` : EMPTY_STRING}${escapeHtml(cell.title)}`;
    const port = `${cell.rangeStart[0]}_${cell.rangeStart[1]}`;

    // A cell is either a single table row (when it has no children), or a table row containing an embedded table (when it has children)
    //          .filter(item => !item.isCollapsed)
    if (cell.children.length === 0) {
      return `     <TR><TD PORT="${port}" ID="${tableId}:${port}" ${styles} ${Dot.cssClassesHref(cell.style.classes)}>${title}</TD></TR>`;
    } else {
      const cellStyles = `BORDER="0"`;
      const tableStyles = styles;

      const dotCell = `     <TR><TD PORT="${port}" ${cellStyles} ${EMPTY_STRING}>${title}</TD></TR>`;

      return `
        <TR><TD BORDER="0" CELLPADDING="0">
        <TABLE ID="${tableId}:${port}" CELLSPACING="8" CELLPADDING="4" CELLBORDER="1" ${tableStyles} BGCOLOR="green" ${Dot.cssClassesHref(cell.style.classes)}>
        ${[dotCell, ...cell.children

          .map(item => Dot.processCell(tableId, item))].join("\n")}
        </TABLE>
        </TD></TR>
      `;
    }
  }

  static processEdges(edges: Edge[]): string {
    return edges
      .map(edge => {
        const from = `${edge.from[0]}:"${edge.from[1]}_${edge.from[2]}"`;
        const to = `${edge.to[0]}:"${edge.to[1]}_${edge.to[2]}"`;

        const attrs = [
          `id="${edge.from[0]}:${edge.from[1]}_${edge.from[2]} -> ${edge.to[0]}:${edge.to[1]}_${edge.to[2]}"`,
          Dot.cssClasses(edge.classes)
        ].filter(s => s !== "").join(", ");

        return `${from} -> ${to} [${attrs}];`;
      })
      .join("\n    ");
  }

  static clusters(subgraphs: Subgraph[]): string {
    return subgraphs
      .map(subgraph => {
        return `
        subgraph "cluster_${subgraph.title}" {
          label = "${subgraph.title}";

          ${subgraph.nodes.join(" ")}

          ${Dot.clusters(subgraph.subgraphs)}
        };
      `;
      })
      .join("\n");
  }

  static cssClasses(classes: CssClass[]): string {
    if (classes.length == 0) {
      return "";
    } else {
      return `class="${Array.from(classes).map(c => c.toString()).join(" ")}"`;
    }
  }

  static cssClassesHref(classes: CssClass[]): string {
    if (classes.length == 0) {
      return "";
    } else {
      return `href="remove_me_url.${Array.from(classes).map(c => c.toString()).join(".")}"`;
    }
  }
}
